src/main/scala/tech/cryptonomic/conseil/routes/Data.scala

```diff
@@ -26,102 +24,125 @@ object Data {
   * @param apiExecutionContext   is used to call the async operations exposed by the api service
   */
 class Data(config: PlatformsConfiguration, queryProtocolPlatform: DataPlatform)(implicit apiExecutionContext: ExecutionContext)
-  extends LazyLogging with RouteHandling with DatabaseApiFiltering with JacksonSupport {
+  extends LazyLogging with DatabaseApiFiltering with DataHelpers {
 
-  import Tezos._
+  import cats.instances.either._
+  import cats.instances.future._
+  import cats.syntax.bitraverse._
+  import cats.instances.option._
+  import cats.syntax.traverse._
 
   /*
    * reuse the same context as the one for ApiOperations calls
    * as long as it doesn't create issues or performance degradation
    */
   override val asyncApiFiltersExecutionContext: ExecutionContext = apiExecutionContext
-
-  /** Route for the POST query */
-  val postRoute: Route =
-    post {
-      commonRoute { (platform, network) =>
-        pathPrefix(Segment) { ent =>
-          validateEntity(ent) {
-            entity(as[ApiQuery]) { apiQuery: ApiQuery =>
-              validateQueryOrBadRequest(ent, apiQuery) { validatedQuery =>
-                completeWithJsonOrNotFound(queryProtocolPlatform.queryWithPredicates(platform, ent, validatedQuery))
-              }
-            }
-          }
+  /** V2 Route implementation for query endpoint */
+  val postRoute: Route = queryEndpoint.implementedByAsync {
+    case ((platform, network, entity), apiQuery, _) =>
+      apiQuery.validate(entity).map { validQuery =>
+        platformNetworkValidation(platform, network) {
+          queryProtocolPlatform.queryWithPredicates(platform, entity, validQuery)
         }
+      }.left.map(Future.successful).bisequence.map(eitherOptionOps)
```

Comment:
I'm a bit lost here, is it trying to expose errors as a result in a Future anyway, when there is something to show?

Reply:
yes, as validation of the query is synchronous and query result is a `Future`, but `implementedByAsync` needs `Future` so I had to do it this way